File: problem8/vers_java/P8.java
/*
 * P8.java
 *
 * Zeke Reyna
 *
 * Reads in a large text file that represents an int, the finds
 * the adjacent digits that form the largest product.
 *
 * In this specific case, the goal is a 100 digit number, and to
 * find the 13 adjacent digits that form the largest product.
 */

import java.util.*;
import java.io.*;

public class P8
{
        public static void print( Object o )
        {
                System.out.print( o );
        }

        public static long find_product( String digits, int adj )
        {
                long product;
                int i;
                char temp;

                product = 1;
                for( i = 0; i < adj; i++ ) {
                        temp = digits.charAt( i );
                        product *= Character.getNumericValue( temp );
                }

                return product;
        }

        public static void main( String args[] )
        {
                int i;
                int adj;
                long largest_product, curr_product;
                String line;

                if( args.length == 1 )
                        adj = Integer.parseInt( args[0] );
                else
                        adj = 4; // default case

                try { //ugh, try catch blocks.
                        BufferedReader br = new BufferedReader( new FileReader("../num.txt"));
                        line = br.readLine();
                        br.close();
                } catch( FileNotFoundException e) {
                        System.out.println( e );
                        return;
                } catch( IOException e ) {
                        System.out.println( e );
                        return;
                }

                largest_product = -1;
                // not so sure about bounds. could subtract 1 from (i + adj)
                for( i = 1; i + adj <= line.length(); i++ ) {

                        curr_product = find_product( line.substring( i, i + adj ), adj );

                        if( curr_product > largest_product )
                                largest_product = curr_product;
                }

                print( "\nThe largest product of the "+adj+" adjacent digits in num.txt" );
                print( " are "+largest_product+"\n\n" );

        }
}

File: problem8/vers_py/p8.py
### p8.py
#
# Zeke Reyna
#
# Reads in a text file and gives the largest product
# of the n adjacent digits in the text.

import sys

def find_product( digits, adj ):

    product = 1
    for i in range( 0, adj ):
        product *= int(digits[ i ])

    return product

def main( argv ):

    if len(argv) == 2:
        adj = int( argv[1] )
    else:
        adj = 4

    f = open( '../num.txt', 'r' )
    number = f.readline() # should be one line long
    f.close()

    largest_product = -1 # bounds?
    for i in range( 0, len(number) - adj ):

        curr_product = find_product( number[i:i+adj], adj )

        if curr_product > largest_product:
            largest_product = curr_product

    print("\nThe largest product of the ", adj, " adjacent digits",
          " in num.txt are ", largest_product, "\n" )


main(sys.argv)
